ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamBaseTest.java

```diff
@@ -390,6 +391,15 @@ protected void runTestCloseStream(List<RaftServer> raftServers, int bufferSize,
     }
   }
 
+  protected void runTestSameStreamId(int numServers, int bufferSize, int bufferNum) throws Exception {
```

Comment:
Let's combine the new test with the existing runTestDataStream(..).  We actually could change back it to check header for all servers (we changed to check only the primary server since call id was unique).
```
+++ b/ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamBaseTest.java
@@ -458,22 +458,27 @@ abstract class DataStreamBaseTest extends BaseTest {
       Assert.assertEquals(reply.getType(), Type.STREAM_DATA);
     }
     try {
-      assertHeader(out.getHeader(), dataSize);
+      for(Server s : servers) {
+        assertHeader(s, out.getHeader(), dataSize);
+      }
     } catch (Throwable e) {
       throw new CompletionException(e);
     }
   }
 
-  void assertHeader(RaftClientRequest header, int dataSize) throws Exception {
-    final Server server = getPrimaryServer();
+  void assertHeader(Server server, RaftClientRequest header, int dataSize) throws Exception {
     final MultiDataStreamStateMachine s = server.getStateMachine(header.getRaftGroupId());
     final SingleDataStream stream = s.getSingleDataStream(header.getCallId());
     final RaftClientRequest writeRequest = stream.getWriteRequest();
     Assert.assertEquals(writeRequest.getCallId(), header.getCallId());
     Assert.assertEquals(writeRequest.getRaftGroupId(), header.getRaftGroupId());
     Assert.assertEquals(raftGroup.getGroupId(), header.getRaftGroupId());
-    Assert.assertEquals(writeRequest.getServerId(), header.getServerId());
     Assert.assertEquals(dataSize, stream.getByteWritten());
+
+    final Server primary = getPrimaryServer();
+    if (server == primary) {
+      Assert.assertEquals(writeRequest.getServerId(), header.getServerId());
+    }
   }
```